torch/_dynamo/output_graph.py

```diff
@@ -259,7 +262,23 @@ def restore_graphstate(self, state: OutputGraphState):
         """Restore a checkpoint created by self.copy_graphstate()"""
         (
             self.graphargs,
-            self.guards,
+            # TODO(voz): Unify checkpointing arch, do checkpointing right
+            # instead of restore graphstate having to write to the context.
+            #
+            # A note on long term direction:
+            # The checkpointing system should know how to restore anything correctly, and not ride on
+            # output graph. Instead, it should be a relatively generic system that can demarcate generations
+            # and do save() to snapshot the current state and increment the generation count, or restore()
+            # to go back to a prior generation. The application of state onto existing structures
+            # should be a detail of registrants with the system, and the system should be vaguely ignorant of
+            # how dependent objects are restored.
+            #
+            # Ex: OutputGraph would register values with the checkpointing system, and when the system calls
+            # restore(), it would do the equivalent of this fn. TracingContext would get hit at the same exact time
+            # with restore(), and so would also restore the guards, allowing dynamo to just make sure the relationship
+            # is preserved, instead of having to handle restoration across systems by brindging from OutputGraph
+            # to TracingContext as we do today.
+            self.tracing_context.guards_context.dynamo_guards,
```

Comment:
It doesn't seem like much work, and would immediately discharge this comment, to make the tracing context responsible for checkpointing itself, rather than teaching into its innards

Reply:
Eh, I was hoping to do checkpointing as the last PR on the stack here, and have it work with shape_env, aot_autograd guards, and dynamo guards.

Reply:
Let me know if you feel strongly about this. The issue around this being more work is not on the restore side, but on getting the layering right for when we checkpoint, on the save side.

Reply:
Eh, I'll do it the simple way now, but eventually I want a real generation tracker abstraction.